File: its-my-training-day/leetcode-top-75/questions/linked-list/heap/merge-k-sorted-lists/solution.cpp
#include "../../../../shared/cpp/ds.hpp"

#include <algorithm>
#include <cctype>
#include <cstdint>
#include <functional>
#include <string>
#include <unordered_map>
#include <vector>

ListNode<int>* mergeKSortedLists(const std::vector<ListNode<int>*>& lists) {
    struct HeapNode {
        int val;
        int idx;
        ListNode<int>* node;
        bool operator<(const HeapNode& other) const { return val < other.val; }
        bool operator>(const HeapNode& other) const { return val > other.val; }
    };
    MinHeap<HeapNode> heap;
    for (int i = 0; i < static_cast<int>(lists.size()); ++i) {
        if (lists[i]) {
            heap.push({lists[i]->val, i, lists[i]});
        }
    }
    ListNode<int> dummy(0);
    ListNode<int>* current = &dummy;
    while (heap.size() > 0) {
        HeapNode top = heap.pop();
        current->next = top.node;
        current = current->next;
        if (top.node->next) {
            heap.push({top.node->next->val, top.idx, top.node->next});
        }
    }
    return dummy.next;
}

File: its-my-training-day/leetcode-top-75/questions/linked-list/heap/merge-k-sorted-lists/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def merge_k_sorted_lists(lists: List[Optional[ListNode]]) -> Optional[ListNode]:
    heap = MinHeap()
    i = 0
    while i < len(lists):
        if lists[i]:
            heap.push((lists[i].val, i, lists[i]))
        i += 1
    dummy = ListNode(0)
    current = dummy
    while heap.size() > 0:
        val, idx, node = heap.pop()
        current.next = node
        current = current.next
        if node.next:
            heap.push((node.next.val, idx, node.next))
    return dummy.next
